### ISSUE TITLE
Parking grid - like vehicle grid, but for parking spaces

### ISSUE THREAD

**Issue by aubergine10** [krzychu124/Cities-Skylines-Traffic-Manager-President-Edition#346]
> #342, #231, #196, #176 and basically everything else that shows up in a search for the term "parking". :)

I think we need some sort of parking grid. Like the vehicle grid, but for parking spaces.

This, primarily, is about optimising pathfinder decision making for mixed-mode transport a.k.a. "park and ride".

Currently pathfinder has no clue how much parking there is at a given spot. Trying to calculate from scratch per PF would be lagmageddon.

Rough idea:

* Map is chopped up in to a grid
    * No idea what ideal cell size would be
* When game loads, the whole grid gets updated (or spread across frames?)
* For each grid cell is total number of parking spaces
* When building or road added/moved/removed, update affected cells accordingly
    * This way there's very little ongoing work
* When a cim tries parking at a given location, if they can't find parking a counter gets added on that grid cell
    * The higher the counter, the less desirable that grid cell is for parking
    * The car can look at nearby cells (and their 'full' counters) to quickly determine where they might find a parking spot, and then try there
    * They can try maximum of ~3 parking cells before giving up and taking some alternate action

It would be nice if there was some way to time-bucket the "parking full" counters, especially if the time buckets could roughly coincide with Real Time mods daily schedule a.k.a. shifts (probably require time sliders so user can choose the bucketing offsets): Early morning, Day, Evening, Night.

The thinking here is that those RT shifts are roughly the time a car would be parked in a given spot. So if someone goes to work their car will be parked for their whole shift.

For users who don't use RT mod, we'd just split day equally in to 4 shifts; 2 daytime, 2 night time. Vanilla game only perceives 2 shifts really: Day and night, and even then the difference between the two is barely noticeable.

So, anyway, the PF can now quickly check a destination to see if it's suitable for parking. If not, it can quickly find the nearest suitable alternative. And if car gets there but can't find parking, a repath is more likely to lead to success.

**Comment by Sipke82** [comment]
Sounds like a cool idea, perhaps make sure #175 ghets fixed first, as now eventually (slowly but surely) all my parking lots are getting filled with "ditched"cars. Likley due to teleporting cims....

**Comment by VictorPhilipp** [comment]
A grid holding parking vehicles aready exists in the base game
![grafik](https://user-images.githubusercontent.com/4333935/58735402-91a3bb00-83fa-11e9-82c9-7caf95f4ffa1.png)
but I am not sure if that suits our needs.

**Comment by aubergine10** [comment]
Yup, if we wanted to do things "reserved parking" as per #231, we would need something more elaborate.
